narwhal/primary/src/tests/block_waiter_tests.rs

```diff
@@ -721,81 +325,64 @@ async fn test_return_error_when_certificate_is_missing_when_get_blocks() {
+    assert_eq!(block_error.error, BlockErrorKind::BlockNotFound);
 }
 
-// worker_listener listens to TCP requests. The worker responds to the
-// RequestBatch requests for the provided expected_batches.
+// Fake worker responds to RequestBatch requests with the provided expected_batches.
 #[must_use]
 pub fn worker_listener(
-    keypair: NetworkKeyPair,
-    address: multiaddr::Multiaddr,
-    expected_batches: HashMap<BatchDigest, BatchMessage>,
-    tx_batch_messages: metered_channel::Sender<BatchResult>,
-) -> JoinHandle<()> {
-    tokio::spawn(async move {
-        let (mut recv, _, _network) = PrimaryToWorkerMockServer::spawn(keypair, address);
-        let mut counter = 0;
-        loop {
-            let message = recv
-                .recv()
-                .await
-                .expect("Failed to receive network message");
-            match message {
-                PrimaryWorkerMessage::RequestBatch(id) => {
-                    if expected_batches.contains_key(&id) {
-                        tx_batch_messages
-                            .send(Ok(expected_batches.get(&id).cloned().unwrap()))
-                            .await
-                            .unwrap();
-
-                        counter += 1;
-
-                        // Once all the expected requests have been received, break the loop
-                        // of the server.
-                        if counter == expected_batches.len() {
-                            break;
-                        }
-                    }
-                }
-                _ => panic!("Unexpected request received"),
-            };
+    worker: &WorkerFixture,
+    expected_batches: HashMap<BatchDigest, Batch>,
+) -> anemo::Network {
+    struct MockPrimaryToWorker {
+        expected_batches: HashMap<BatchDigest, Batch>,
+    }
+    #[async_trait]
+    impl PrimaryToWorker for MockPrimaryToWorker {
+        async fn send_message(
+            &self,
+            _request: anemo::Request<PrimaryWorkerMessage>,
+        ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
+            unimplemented!()
         }
-    })
+        async fn synchronize(
+            &self,
+            _request: anemo::Request<WorkerSynchronizeMessage>,
+        ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
+            unimplemented!()
+        }
+        async fn request_batch(
+            &self,
+            request: anemo::Request<RequestBatchRequest>,
+        ) -> Result<anemo::Response<RequestBatchResponse>, anemo::rpc::Status> {
+            Ok(anemo::Response::new(RequestBatchResponse {
+                batch: self.expected_batches.get(&request.body().batch).cloned(),
+            }))
+        }
+        async fn delete_batches(
+            &self,
+            _request: anemo::Request<WorkerDeleteBatchesMessage>,
+        ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
+            unimplemented!()
+        }
+    }
```

Comment:
I suppose we can leave trying mockall for this in a later PR (given this is already a large PR)

Reply:
Yes that was my plan!